src/core/server/saved_objects/migrations/model/helpers.ts

```diff
@@ -68,6 +69,55 @@ export function indexBelongsToLaterVersion(indexName: string, kibanaVersion: str
   return version != null ? gt(version, kibanaVersion) : false;
 }
 
+export function addMustNotClausesToQuery(
+  query: QueryDslQueryContainer,
+  mustNotClauses: QueryDslQueryContainer[]
+): QueryDslQueryContainer {
+  let filter: QueryDslQueryContainer[] | QueryDslQueryContainer = [];
+
+  if (query.bool?.filter) {
+    if (Array.isArray(query.bool.filter)) {
+      filter = filter.concat(query.bool.filter);
+
+      filter.push({
+        bool: {
+          must_not: mustNotClauses,
+        },
+      });
+    } else {
+      let currentMustNot: QueryDslQueryContainer[] = [];
+
+      if (query.bool.filter.bool?.must_not) {
+        currentMustNot = currentMustNot.concat(query.bool.filter.bool.must_not);
+      }
+
+      currentMustNot.push(...mustNotClauses);
+
+      filter = {
+        ...query.bool.filter,
+        bool: {
+          ...query.bool.filter.bool,
+          must_not: currentMustNot,
+        },
+      };
```

Comment:
I feel this is introducing some unnecessary complexity. I would just handle this case in a similar way than when the existing filter is an array. I think we can even merge the two blocks:

```ts
filter = [
   ...(Array.isArray(query.bool.filter) ? query.bool.filter : [query.bool.filter]),  
   {
      bool: {
        must_not: mustNotClauses,
      },
   }
]
``` 

WDYT?

Reply:
I wanted to simplify the end query, but I guess simplifying the code is a smarter move in this scenario (as long as there's no difference in terms of performance).

Reply:
I don't think there will be any perf concern (and even if, I would prefer to have a more consistent terms aggregation logic with homogeneous perf I think)

Reply:
I didn't fully investigate this, so maybe it won't work, but I wonder if we can't just collect all the `must_not`'s in the model and only build the query in the action.

If the state just has `excludeOnUpgradeMustNots: Array<QueryDslQueryContainer>` then any step in the migration could just concat its must_nots to the state with very straight forward logic.

Then in the end it's very straightforward to wrap the must_nots to create the final query.

We've got good code coverage to mitigate it, but still I feel like this helper is rather complex and a potential source of bugs which might be hard to understand.

Reply:
Alright, I just greatly simplified the helpers to a minimal expression in my latest commit.